File: CSM_Exemple_5_PI_1r_ordre.m
%%% Exemple sintonització analítica PI de 1r ordre
% Identificar el model, el diagrama de Bode, i implementar PI
% J.A.Mayugo, UdG, 2022

clear; close all;

%% Propietats del sistema
U  = 5;         % V, input 'bump test'
omega_ss= 104;  % rad/s
t0 = 300e-3;    % s
t1 = 375e-3;    % s

omega_n = 25;   % rad/s
zeta    = 1;    % críticament esmorteït

s = tf('s');
t_ = 0:0.001:0.6;    % resposta temporal

%% APARTAT a)
tau = t1-t0;
K = omega_ss/U;

G_m = K/(tau*s+1);

% Resposta en llaç obert:
U_ = U *ones(size(t_));
[omg,T,X]         = lsim(G_m,U_,t_);  % voltage/angular vel.
H_ol=figure;
plot(t_,omg,'LineWidth',1.8)
grid; title(['Resposta a una funció escaló amb ' num2str(U) ' V al motor'],'FontSize',24)
xlabel('$t$ [s]','Interpreter','latex','FontSize',24);ylabel('$\omega$ [rad/s]','Interpreter','latex','FontSize',24)
set(gca,'FontSize',18)
saveas(H_ol,'./figures/Exemple_PI_1r_ordre_ol.pdf')

% FUNCIÓ AMB L'ORDRE BODE de MatLab i R-Locus
H_bode=figure
h = bodeplot(G_m); %,{0.1,20});
%setoptions(h,'MagUnits','abs','FreqScale','linear')

H_rlocus=figure;
rlocus(G_m);
p = pole(G_m)

%% APARTATS b) i c) Sintonització analítica d'un PI
% control velocitat llaç tancat considerant un sistema de 2n ordre
Kp = (2*zeta*omega_n*tau-1)/K 
Ki = omega_n^2*tau/K
PI = Kp + Ki/s              % controlador PI
G_cl = PI*G_m/(1+PI*G_m)    % model llaç tancat amb PI, fórmula de Mason
p_cl = pole(G_cl)

u = omega_ss;
u_ = u *ones(size(t_));
[omg_,T,X] = lsim(G_cl,u_,t_);  % voltage/angular vel.
H_cl=figure;
plot(t_,omg_,'LineWidth',1.8)
grid; title(['Resposta control PI, consigna ' num2str(u) ' rad/s'],'FontSize',24)
xlabel('$t$ [s]','Interpreter','latex','FontSize',24);ylabel('$\omega$ [rad/s]','Interpreter','latex','FontSize',24)
legend(["$K_p$ = " + num2str(round(Kp,2)) + ...
        ", $K_i$ = " + num2str(round(Ki,2))], ...
       'Interpreter','latex','FontSize',18,'Location','south');
ylim([0,120]);set(gca,'FontSize',18);legend boxoff
saveas(H_cl,'./figures/Exemple_PI_1r_ordre_cl.pdf')

%% APARTAT d) Error de variació de sortida: 'Error de velocitat'

e_v=1/(Ki*K); % rad/s2, error en la derivada de del valor de sortida

disp(["L'error estacionari és de "+num2str(round(e_v,4))+'rad/s2'])

%% Alternativa) Direct Syntesis Method
% control velocitat llaç tancat controlador PI
lambda_ = [10, 20, 50, 100];
u = omega_ss;
u_ = u *ones(size(t_));
for ii = 1:length(lambda_)
    Kp_(ii)= tau/K*lambda_(ii);
    Ti     = tau;
    
    PI2   = Kp_(ii)*(Ti*s+1)/(Ti*s); % controlador PI
    PI2   = Kp_(ii)*(1+1/(Ti*s)); % controlador PI
    G2_cl = PI2*G_m/(1+PI2*G_m);% model llaç tancat amb PI, fórmula de Mason
    p2_cl = pole(G2_cl);
    
    [omg_,T,X] = lsim(G2_cl,u_,t_);  % voltage/angular vel.
    OMG_(ii,:)=omg_;
end

H2_cl=figure;
hold on
legend_=string(lambda_);
for ii = 1:length(lambda_)
    plot(t_,OMG_(ii,:),'LineWidth',1.8);
    legend_(ii) = '$\lambda$ = ' + legend_(ii) + ', $K_p$ = ' + ...
                  num2str(round(Kp_(ii),2)) + ', $K_i$ = ' + ...
                  num2str(round(Kp_(ii)/Ti,2));
end
grid; title(['Resposta control PI, consigna ' num2str(u) ' rad/s'],'FontSize',24)
xlabel('$t$ [s]','Interpreter','latex','FontSize',24);
ylabel('$\omega$ [rad/s]','Interpreter','latex','FontSize',24);
legend(legend_,'Interpreter','latex','FontSize',18,'Location','south');
ylim([0,120]);set(gca,'FontSize',18);legend boxoff
saveas(H2_cl,'./figures/Exemple_PI_1r_ordre_2cl.pdf')

File: CSM_Exemple_5_PI_1r_ordre.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
%% Exemple 5 PI 1r ordre
%  J.A.Mayugo, UdG, 2023
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy import signal


# %% Propietats del sistema
U  = 5;         # % V, input 'bump test'
omega_ss= 104;  # % rad/s
t0 = 300e-3;    # % s
t1 = 375e-3;    # % s

omega_n = 25;   # % rad/s
zeta    = 1;    # % críticament esmorteït

# %% Resolució APARTAT a)
tau = t1-t0;
K = omega_ss/U;

# % Model
numG = [K]
denG = [tau, 1]
tf_G_m = signal.lti(numG, denG)

# % Resposta transitòria: escaló en llaç obert:

tRange = np.linspace(0, 0.6, num=1000)

t, y = tf_G_m.step(T=tRange)
# t, y = signal.step2(tf_G_m,T=tRange)
plt.plot(t, y*U)
plt.xlabel('t [s]')
plt.ylabel('$\omega_m$ [rad/s]')
plt.title('Resposta a una funció escaló amb ' + str(U) + 'V al motor')
plt.xlim([0, t[-1]])
plt.ylim([0, 110])
plt.grid()

# % Diagrama de Bode i R-Locus
wRange = np.logspace(np.log10(0.1), np.log10(1e4), num=100)
w, mag, phase = tf_G_m.bode(w=wRange)

plt.subplots(2, 1, sharex='col')
plt.subplot(2, 1, 1)
plt.semilogx(w, mag)    # Bode magnitude plot
plt.title('Bode Diagram: input: Volts, output: omega_m')
plt.ylabel('Magnitude [Db]')
plt.subplot(2, 1, 2)
plt.semilogx(w, phase)  # Bode phase plot
plt.ylabel('Phase [deg]')
plt.xlabel('Frequency [rad/s]')

w, H = tf_G_m.freqresp()
plt.figure()
plt.plot(H.real, H.imag, "b")
plt.plot(H.real, -H.imag, "r")
plt.xlabel('Real Axix [1/s]')
plt.ylabel('Imaginary Axix [1/s]')
plt.title('R-Locus: input: Volts, output: omega_m')

print('Pols del sistema G_m')
print(tf_G_m.poles)

# %% Resolució APARTAT b i c) Sintonització analítica d'un PI
# % control velocitat llaç tancat considerant un sistema de 2n ordre

Kp = (2*zeta*omega_n*tau-1)/K 
Ki = omega_n**2*tau/K

print(f'Contrololador PID, paràmetre K_p {Kp:.4f}')
print(f'Contrololador PID, paràmetre K_i {Ki:.4f}')

# PI = Kp + Ki/s             # % controlador PI
numC = [Kp, Ki]
denC = [1, 0]

num_ol = np.polymul(numC, numG)
den_ol = np.polymul(denC, denG)

# feedback manual: T = N / (D + N)
    # model_G_cl = PI*model_G_m/(1+PI*model_G_m)   # % model llaç tancat amb PI, fórmula de Mason
num_cl = num_ol
den_cl = np.polyadd(den_ol, num_ol)

tf_G_cl = signal.lti(num_cl, den_cl)   

print('Pols del sistema G_cl')
print(tf_G_cl.poles)

# %%

t, y = tf_G_cl.step(T=tRange)
# t, y = signal.step2(tf_G_cl,T=tRange)

plt.figure()
plt.plot(t, y*omega_ss)
plt.xlabel('t [s]')
plt.ylabel('$\omega_m$ [rad/s]')
plt.title('Resposta control PI, consigna ' + str(omega_ss) + ' rad/s')
plt.grid()
plt.legend(["$K_p$ = " + str(round(Kp,2)) + ", $K_i$ = " + str(round(Ki,2))])
plt.xlim([0, t[-1]])
plt.ylim([0, 110])

# %% Resolució APARTAT d) Error de variació de sortida: 'Error de velocitat'

e_v=1/(Ki*K)  #% rad/s2, error en la derivada de del valor de sortida

print(["L'error estacionari és de velocitat és "+ str(round(e_v,4))+' rad/s2'])


# %% Alternativa) Direct Syntesis Method
# % control velocitat llaç tancat controlador PI
lambda_ = np.array([10, 20, 50, 100])
u = omega_ss;
Kp_ = np.zeros(lambda_.shape[0])
y_  = np.zeros([lambda_.shape[0],tRange.shape[0]])

for ii in range(lambda_.shape[0]):

    Kp_[ii] = tau/K * lambda_[ii]
    Ti = tau
    
    numC = [Kp_[ii]*Ti, Kp_[ii]]        # PI: Kp*(1 + 1/(Ti*s))
    denC = [Ti, 0]

    # llaç obert
    num_ol = np.polymul(numC, numG)
    den_ol = np.polymul(denC, denG)

    # llaç tancat: T = N / (D + N)
    num_cl = num_ol
    den_cl = np.polyadd(den_ol, num_ol)

    sys_cl = signal.lti(num_cl, den_cl)

    t, y = signal.step(sys_cl, T=tRange)
    y_[ii, :] = y

plt.figure()
# legend_=string(lambda_);
for ii in range(0,lambda_.shape[0]):
    plt.plot(t, y_[ii,:]*omega_ss,
             label='$\lambda$ = ' + str(round(lambda_[ii],0)) + 
                   ", $K_p$ = " + str(round(Kp_[ii],2)) + 
                   ", $T_i$ = " + str(round(Ti,3)) + 
                   ' ($K_i$ = '  + str(round(Kp_[ii]/Ti,2)) +')' ) 

plt.title('Resposta control PI, consigna ' + str(omega_ss) + ' rad/s')
plt.grid()
plt.xlabel('t [s]')
plt.ylabel('$\omega_m$ [rad/s]')
plt.legend()
plt.xlim([0, t[-1]])
plt.ylim([0, 110])

plt.show()